Consistently use unsigned/size_t in big-int [blocks: #2310]

tautschnig opened the pull request.

tautschnig committed:
Consistently use unsigned in big-int

tautschnig committed:
Consistently use unsigned in big-int

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: a323dd22a5d5aad574c5555337ffb26c99aba0db).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77183230
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).
##

kroening commented:
The index shifts are certainly scary.

Equally scary is the idea that most of these 'unsigned' should likely be size_t, even though that means that we worry about an integer that needs more than 2^32 bits to represent.

allredj commented:
Could we have unit tests for these, to make sure that we're not breaking any corner case?

tautschnig commented:
I have added a commit to move to `size_t` and have also enabled a unit test for `BigInt`.

tautschnig committed:
Use static_cast instead of C-style cast

tautschnig committed:
Use static_cast instead of C-style cast

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e5501eb35355a34d301e3fd29090dab9af933a12).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91047441

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d63c0495ce4bd7575d02a566c74a010d0343345c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/91344288

NlightNFotis commented:
Hi @tautschnig, is this something you plan on merging soon?

If you'd like it merged but don't have the time to work on this, is there something I can do to get this in a better shape to merge?

TGWDB commented:
Closing due to age (no further comment on PR content), please reopen with rebase on develop if you intent to continue this work.

tautschnig committed:
Use std::size_t for all size-typed operations in big-int

tautschnig committed:
Use std::size_t for all size-typed operations in big-int

tautschnig commented:
@kroening This has been rebased and I hope to have addressed all past concerns.

tautschnig committed:
Avoid signed overflow when negating llong_t values in big-int

Negating the most-negative signed value (LLONG_MIN) is undefined
behavior in C++. Fix by casting to unsigned first, then negating:
-static_cast<ullong_t>(v) is well-defined for all signed values.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>

tautschnig committed:
Avoid signed overflow when negating llong_t values in big-int

Negating the most-negative signed value (LLONG_MIN) is undefined
behavior in C++. Fix by casting to unsigned first, then negating:
-static_cast<ullong_t>(v) is well-defined for all signed values.

Co-authored-by: Kiro <kiro-agent@users.noreply.github.com>
